README.md
# create-servers

Create an http AND/OR an https server and call the same request handler.

## Usage

The `create-servers` module exports a function that takes a config object and
a node-style callback. The config object must have at minimum an `http` or
`https` property (or both). The following config properties are supported:

| Property                 | Description                                                                                                                                                                                                                                                                           |
| ------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `handler`                | Request handler to be used for any server, unless overridden specifically with `http.handler` or `https.handler`.                                                                                                                                                                     |
| `timeout`                | Socket timeout in milliseconds for any server, unless overridden with `http.timeout` or `https.timeout`. Defaults to the node default of 2 minutes.                                                                                                                                   |
| `keepAliveTimeout`       | Milliseconds of activity before sockets are destroyed. Defaults to the node default value (currently 5 seconds). |
| `http`                   | Optional. If present, an HTTP server is started. This can be an object or a number. If it's a number, it's used as the TCP port for an HTTP server. You may also use an Array to start multiple servers.                                                                              |
| `http.port`              | TCP port for the HTTP server. Defaults to `80`.                                                                                                                                                                                                                                       |
| `http.host`              | The address the HTTP server is bound to. Defaults to `::` or `0.0.0.0`.                                                                                                                                                                                                               |
| `http.timeout`           | Socket timeout in milliseconds for the server. If unspecified, the top-level `timeout` configuration is used.                                                                                                                                                                         |
| `http.keepAliveTimeout`  | Overrides the top-level keepAliveTimeout setting if specified. |
| `http.handler`           | Handler for HTTP requests. If you want to share a handler with all servers, use a top-level `handler` config property instead.                                                                                                                                                        |
| `https`                  | Optional object. If present, an HTTPS server is started. You may start multiple HTTPS servers by passing an array of objects                                                                                                                                                          |
| `https.port`             | TCP port for the HTTPS server. Defaults to `443`.                                                                                                                                                                                                                                     |
| `https.host`             | The address the HTTPS server is bound to. Defaults to `::` or `0.0.0.0`.                                                                                                                                                                                                              |
| `https.timeout`          | Socket timeout in milliseconds for the server. If unspecified, the top-level `timeout` configuration is used.                                                                                                                                                                         |
| `https.keepAliveTimeout`  | Overrides the top-level keepAliveTimeout setting if specified. |
| `https.ciphers`          | Defaults to a [default cipher suite](#note-on-security). To customize, either supply a colon-separated string or array of strings for the ciphers you want the server to support.                                                                                                     |
| `https.honorCipherOrder` | If true, prefer the server's specified cipher order instead of the client's. Defaults to `false`.                                                                                                                                                                                     |
| `https.root`             | Root directory for certificate/key files. See [Certificate normalization](#certificate-normalization) for more details.                                                                                                                                                               |
| `https.key`              | PEM/file path for the server's private key. See [Certificate normalization](#certificate-normalization) for more details.                                                                                                                                                             |
| `https.cert`             | PEM/file path(s) for the server's certificate. See [Certificate normalization](#certificate-normalization) for more details.                                                                                                                                                          |
| `https.ca`               | Cert or array of certs specifying trusted authorities for peer certificates. Only required if your server accepts client certificate connections signed by authorities that are not trusted by default. See [Certificate normalization](#certificate-normalization) for more details. |
| `https.sni`              | See [SNI Support](#sni-support).                                                                                                                                                                                                                                                      |
| `https.handler`          | Handler for HTTPS requests. If you want to share a handler with all servers, use a top-level `handler` config property instead.                                                                                                                                                       |
| `https.*`                | Any other properties supported by [https.createServer](https://nodejs.org/dist/latest-v8.x/docs/api/https.html#https_https_createserver_options_requestlistener) can be added to the https object, except `secureProtocol` and `secureOptions` which are set to recommended values.   |
| `http2`                  | Optional object. If present, an HTTP/2 server is started. You may start multiple HTTP/2 servers by passing an array of objects                                                                                                                                                        |
| `http2.allowHTTP1`       | Enable [ALPN negotiation] allowing support for both HTTPS and HTTP/2 on the same socket.                                                                                                                                                                                              |
| `http2.*`                | The same `https` security options are allowed, as well as any other properties supported by [http2.createSecureServer](https://nodejs.org/dist/latest-v8.x/docs/api/http2.html#http2_http2_createsecureserver_options_onrequesthandler).                                              |

If successful, the `create-servers` callback is passed an object with the
following properties:

| Property | Description                                                                                     |
| -------- | ----------------------------------------------------------------------------------------------- |
| `http`   | The HTTP server that was created, if any. If creating multiple servers, this will be an Array.  |
| `https`  | The HTTPS server that was created, if any. If creating multiple servers, this will be an Array. |

### Certificate Normalization

`create-servers` provides some conveniences for `https.ca`, `https.key`, and
`https.cert` config properties. You may use PEM data directly (inside a `Buffer`
or string) or a file name. When using a file name, you must also set an
`https.root` config property if using relative paths to cert/key files.

`https.ca`, `https.cert`, and `https.key` also support specifying an Array.
Given an array for `cert`, you must have a matching array for `key` so each cert
can be matched with its private key.

```js
const createServers = require('create-servers');

createServers({
  https: {
    root: '/cert/path',
    cert: ['cert1.crt', 'cert2.crt'],
    key: ['cert1.key', 'cert2.key']
  }
}, err => {
  // ...
})
```

If you have a cert that is signed by an intermediate CA, your server will need
to append the untrusted parts of the CA chain with your cert. To make this more
convenient, `create-servers` lets you use an array to automatically create a
chain.

```js
const createServers = require('create-servers');

createServers({
  https: {
    root: '/cert/path',
    cert: ['cert.crt', 'intermediate.crt'],
    key: 'cert.key'
  }
}, err => {
  // ...
})
```

If you are specifying multiple certs _and_ you want to create chains for each,
use an array of arrays.

```js
const createServers = require('create-servers');

createServers({
  https: {
    root: '/cert/path',
    cert: [['cert1.crt', 'intermediate.crt'], 'cert2.crt'],
    key: ['cert1.key', 'cert2.key']
  }
}, err => {
  // ...
})
```

### SNI Support

[Server Name Indication](https://en.wikipedia.org/wiki/Server_Name_Indication),
or SNI, lets HTTPS clients announce which hostname they wish to connect to
before the server sends its certificate, enabling the use of the same server for
multiple hosts. Although `SNICallback` can be used to support this, you lose the
convenient certificate normalization provided by `create-servers`. The `sni`
config option provides an easier way.

The `sni` option is an object with each key being a supported hostname and each
value being a subset of the HTTPS settings listed above. HTTPS settings defined
at the top level are used as defaults for the hostname-specific settings.

```js
const createServers = require('create-servers');

createServers(
  {
    https: {
      port: 443,
      sni: {
        'example1.com': {
          key: '/certs/private/example1.com.key',
          cert: '/certs/public/example1.com.crt'
        },
        'example2.com': {
          key: '/certs/private/example2.com.key',
          cert: '/certs/public/example2.com.crt'
        }
      }
    },
    handler: function (req, res) {
      res.end('Hello');
    }
  },
  function (errs) {
    if (errs) {
      return console.log(errs.https);
    }

    console.log('Listening on 443');
  }
);
```

Use `*` in the hostname for wildcard certs. Example: `*.example.com`. The
following settings are supported in the host-specific configuration:

* key
* cert
* ca
* ciphers
* honorCipherOrder
* Anything else supported by [`tls.createSecureContext`](https://nodejs.org/dist/latest-v8.x/docs/api/tls.html#tls_tls_createsecurecontext_options)

## NOTE on Security
Inspired by [`iojs`][iojs] and a well written [article][article], we have defaulted
our [ciphers][ciphers] to support "perfect-forward-security" as well as removing insecure
cipher suites from being a possible choice. With this in mind,
be aware that we will no longer support ie6 on windows XP by default.

## Examples

### http

```js
var createServers = require('create-servers');

var servers = createServers(
  {
    http: 80,
    handler: function (req, res) {
      res.end('http only');
    }
  },
  function (errs) {
    if (errs) {
      return console.log(errs.http);
    }

    console.log('Listening on 80');
  }
);
```

### https

```js
var servers = createServers(
  {
    https: {
      port: 443,
      root: '/path/to/ssl/files',
      key: 'your-key.pem',
      cert: 'your-cert.pem',
      ca: 'your-ca.pem' // Can be an Array of CAs
    },
    handler: function (req, res) {
      res.end('https only');
    }
  },
  function (errs) {
    if (errs) {
      return console.log(errs.https);
    }

    console.log('Listening on 443');
  }
);
```

### http && https

```js
var servers = createServers(
  {
    http: 80,
    https: {
      port: 443,
      root: '/path/to/ssl/files',
      key: 'your-key.pem',
      cert: 'your-cert.pem',
      ca: 'your-ca.pem' // Can be an Array of CAs
    },
    handler: function (req, res) {
      res.end('http AND https');
    }
  },
  function (errs, servers) {
    if (errs) {
      return Object.keys(errs).forEach(function (key) {
        console.log('Error ' + key + ': ' + errs[key]);
        if (servers[key]) {
          servers[key].close();
        }
      });
    }

    console.log('Listening on 80 and 443');
  }
);
```

### http && https (different handlers)

```js
var servers = createServers(
  {
    http: {
      port: 80,
      handler: function (req, res) {
        res.end('http');
      }
    },
    https: {
      port: 443,
      root: '/path/to/ssl/files',
      key: 'your-key.pem',
      cert: 'your-cert.pem',
      ca: 'your-ca.pem', // Can be an Array of CAs
      handler: function (req, res) {
        res.end('https');
      }
    }
  },
  function (errs, servers) {
    if (errs) {
      return Object.keys(errs).forEach(function (key) {
        console.log('Error ' + key + ': ' + errs[key]);
        if (servers[key]) {
          servers[key].close();
        }
      });
    }

    console.log('Listening on 80 and 443');
  }
);
```

## Author: [Charlie Robbins](https://github.com/indexzero)
## License: MIT

[article]: https://certsimple.com/blog/a-plus-node-js-ssl
[iojs]: https://github.com/iojs/io.js
[ciphers]: https://iojs.org/api/tls.html#tls_tls_createserver_options_secureconnectionlistener
[ALPN negotiation]: https://nodejs.org/api/http2.html#http2_alpn_negotiation


CHANGELOG.md
# CHANGELOG

## 3.3.0

- Support `keepAliveTimeout`

## 3.2.1

- [#50] Fix bug where cert file reading errors do not surface

## 3.2.0

- [#37] Support HTTP2

## 3.1.0

- [#23], @DullReferenceException Support "*" for SNI host pattern

## 3.0.1

- [#22] Return value compatibility. Do no return `null` valued keys.

## 3.0.0

- [#21], @DullReferenceException Support creating multiple HTTP or HTTPS servers. 
  - [BREAKING] Introduces `async` functions. 


index.d.ts
import { Server as HttpServer } from 'http';
import { Server as HttpsServer, ServerOptions as NodeHttpsServerOptions } from 'https';
import { Http2SecureServer, SecureServerOptions as NodeHttp2ServerOptions } from 'http2';

type IPAddress = string;
type FilePath = string;

type PEMPath = string;
type PEMData = string | Buffer;
type PEM = PEMPath | PEMData;
type PEMChain = PEM[];
type PEMs = PEM | PEMChain;
type CertificatePEMs = PEMs | PEMs[];

type CommonOptions = {
  timeout?: number,
  handler?: Function,
}

type ServerOptions = CommonOptions & {
  port?: number,
  host: IPAddress,
}

type HttpOptions = ServerOptions;

type TLSOptions = {
  ciphers?: string,
  honorCipherOrder?: boolean,
  /**
   * Required if ca, cert or key are FilePath
   */
  root?: FilePath,
  key?: PEM,
  cert?: CertificatePEMs,
  ca?: PEMs,
}

type SNIOptions = {
  sni?: {
    [hostname: string]: TLSOptions
  }
}

type HttpsOptions = ServerOptions & TLSOptions & SNIOptions & NodeHttpsServerOptions;

type Http2Options = HttpsOptions & NodeHttp2ServerOptions;

type Options = {
  http?: HttpOptions | Array<HttpOptions>,
  https?: HttpsOptions | Array<HttpsOptions>,
  http2?: Http2Options | Array<Http2Options>,
  handler?: Function,
}

type Errors = {
  message: string,
  http?: Error | Error[],
  https?: Error | Error[],
  http2?: Error | Error[],
}

type Servers = {
  http?: HttpServer | HttpServer[],
  https?: HttpsServer | HttpsServer[],
  http2?: Http2SecureServer | Http2SecureServer[],
}

type Callback = (error?: Errors, servers?: Servers) => void;

declare function createServers(options: Options, callback: Callback): void;

export = createServers;


index.js
'use strict';

/*
 * index.js: Create an http AND/OR an https server and call the same request handler.
 *
 * (C) 2013, Charlie Robbins.
 *
 */

const
  fs = require('fs').promises,
  tls = require('tls'),
  path = require('path'),
  constants = require('constants'),
  connected = require('connected'),
  errs = require('errs'),
  assign = require('object-assign');

const pemFormat = /-----BEGIN/;

const CIPHERS = [
  'ECDHE-RSA-AES256-SHA384',
  'DHE-RSA-AES256-SHA384',
  'ECDHE-RSA-AES256-SHA256',
  'DHE-RSA-AES256-SHA256',
  'ECDHE-RSA-AES128-SHA256',
  'DHE-RSA-AES128-SHA256',
  'HIGH',
  '!aNULL',
  '!eNULL',
  '!EXPORT',
  '!DES',
  '!RC4',
  '!MD5',
  '!PSK',
  '!SRP',
  '!CAMELLIA'
].join(':');

const secureOptions = constants.SSL_OP_NO_SSLv3;

/**
 * function createServers (dispatch, options, callback)
 * Creates and listens on both HTTP and HTTPS servers.
 */
module.exports = async function createServers(options, listening) {
  try {
    options = normalizeOptions(options);
  } catch (err) {
    return listening(err);
  }

  const [httpResult, httpsResult, http2Result] = await Promise.allSettled([
    createHttp(options.http, options.log),
    createHttps(options.https, options.log),
    createHttps(options.http2, options.log, true)
  ])

  const servers = {};
  if (httpResult.value) servers.http = httpResult.value;
  if (httpsResult.value) servers.https = httpsResult.value;
  if (http2Result.value) servers.http2 = http2Result.value;

  const errorSource = httpResult.reason || httpsResult.reason || http2Result.reason;
  if (errorSource) {
    if (Array.isArray(errorSource)) {
      errorSource = errorSource[0];
    }
    return listening(
      errs.create({
        message: errorSource && errorSource.message,
        http2: http2Result.reason,
        https: httpsResult.reason,
        http: httpResult.reason
      }),
      servers
    );
  }

  listening(undefined, servers);
};

function normalizeOptions(options) {
  const http = normalizeHttpOptions(options.http, options);
  const https = normalizeHttpsOptions(options.https, options);
  const http2 = normalizeHttpsOptions(options.http2, options);

  if (!http && !https && !http2) {
    throw new Error('http, https, and/or http2 are required options');
  }

  return {
    http,
    https,
    http2,
    log: options.log || function() {}
  };
}

function normalizeHttpOptions(httpConfig, baseConfig) {
  if (typeof httpConfig === 'undefined') return;

  if (Array.isArray(httpConfig)) {
    return httpConfig.map(cfg => normalizeHttpOptions(cfg, baseConfig));
  }

  let port =
    typeof httpConfig === 'object' && 'port' in httpConfig
      ? httpConfig.port
      : httpConfig;
  if (typeof port === 'undefined') {
    port = 80;
  }

  const http = {
    host: httpConfig.host || baseConfig.host,
    port: +port,
    handler: httpConfig.handler || baseConfig.handler,
    timeout: httpConfig.timeout || baseConfig.timeout,
    keepAliveTimeout: httpConfig.keepAliveTimeout || baseConfig.keepAliveTimeout
  };

  if (!http.handler) {
    throw new Error('handler option is required');
  }

  return http;
}

function normalizeHttpsOptions(httpsConfig, baseConfig) {
  if (typeof httpsConfig === 'undefined') return;

  if (Array.isArray(httpsConfig)) {
    return httpsConfig.map(cfg => normalizeHttpsOptions(cfg, baseConfig));
  }

  const https = {
    ...httpsConfig,
    host: httpsConfig.host || baseConfig.host,
    port: +('port' in httpsConfig ? httpsConfig.port : 443),
    handler: httpsConfig.handler || baseConfig.handler,
    timeout: httpsConfig.timeout || baseConfig.timeout,
    keepAliveTimeout: httpsConfig.keepAliveTimeout || baseConfig.keepAliveTimeout
  };

  if (!https.handler) {
    throw new Error('handler option is required');
  }

  return https;
}

function normalizeCertContent(root, cert, key) {
  // Node accepts an array of certs, which must match up with an array of keys.
  // The user may instead intend for an array passed into cert to represent
  // a cert chain they want to concatenate. Therefore, if key is not an array,
  // we'll assume the latter.
  if (Array.isArray(cert)) {
    if (Array.isArray(key)) {
      // This is an array of certs/chains with corresponding keys
      return normalizeCertChainList(root, cert);
    } else {
      // This is a single cert chain
      return normalizeCertChain(root, cert);
    }
  }

  return normalizePEMContent(root, cert);
}

function normalizeCertChainList(root, data) {
  // If this is an array, treat like an array of bundles, otherwise a single
  // bundle
  return Array.isArray(data)
    ? Promise.all(data.map(function(item) {
        return normalizeCertChain(root, item);
      }))
    : normalizePEMContent(root, data);
}

async function normalizeCertChain(root, data) {
  // A chain can be an array, which we concatenate together into one PEM,
  // an already-concatenated chain, or a single PEM

  const content = await normalizePEMContent(root, data);
  return Array.isArray(content) ? content.join('\n') : content;
}

function normalizeCA(root, ca) {
  if (ca && !Array.isArray(ca)) {
    ca = [ca];
  }
  return ca && Promise.all(ca.map(normalizePEMContent.bind(null, root)));
}

/**
 * function normalizePEMContent(root, file)
 * Returns the contents of `file` verbatim if it is determined to be
 * certificate material and not a file path. Otherwise, returns the
 * certificate material read from that file path.
 */
function normalizePEMContent(root, file) {
  if (Array.isArray(file))
    return Promise.all(file.map(function map(item) {
      return normalizePEMContent(root, item);
    }));

  //
  // Assumption that this is a Buffer, a PEM file, or something broken
  //
  if (typeof file !== 'string' || pemFormat.test(file)) {
    return file;
  }

  return fs.readFile(path.resolve(root, file));
}

function normalizeCiphers(ciphers) {
  ciphers = ciphers || CIPHERS;
  //
  // Remark: If an array is passed in lets join it like we do the defaults
  //
  if (Array.isArray(ciphers)) {
    ciphers = ciphers.join(':');
  }
  return ciphers;
}

async function getSNIHandler(sslOpts) {
  const sniHosts = Object.keys(sslOpts.sni);

  // Pre-compile regexps for the hostname
  const hostRegexps = sniHosts.map(function(host) {
    return host === '*' ? /.*/ : new RegExp(
      '^' +
      host
        .replace('.', '\\.') // Match dots, not wildcards
        .replace('*\\.', '(?:.*\\.)?') + // Handle optional wildcard sub-domains
        '$',
      'i'
    );
  });

  // Prepare secure contexts ahead-of-time
  const hostSecureContexts = await Promise.all(sniHosts.map(async function(host) {
    var hostOpts = sslOpts.sni[host];

    var root = hostOpts.root || sslOpts.root;

    const [key, cert, ca] = await Promise.all([
      normalizePEMContent(root, hostOpts.key),
      normalizeCertContent(root, hostOpts.cert),
      normalizeCA(root, hostOpts.ca || sslOpts.ca)
    ])

    return tls.createSecureContext(
      assign({}, sslOpts, hostOpts, {
        key,
        cert,
        ca,
        ciphers: normalizeCiphers(hostOpts.ciphers || sslOpts.ciphers),
        honorCipherOrder: !!(
          hostOpts.honorCipherOrder || sslOpts.honorCipherOrder
        ),
        secureProtocol: 'SSLv23_method',
        secureOptions: secureOptions
      })
    );
  }));

  return function(hostname, cb) {
    var matchingHostIdx = sniHosts.findIndex(function(candidate, i) {
      return hostRegexps[i].test(hostname);
    });

    if (matchingHostIdx === -1) {
      return void cb(new Error('Unrecognized hostname: ' + hostname));
    }

    cb(null, hostSecureContexts[matchingHostIdx]);
  };
}

//
// ### function createHttp (httpConfig)
// Attempts to create and listen on the the HTTP server.
//
async function createHttp(httpConfig, log) {
  if (typeof httpConfig === 'undefined') {
    log('http | no options.http; no server');
    return null;
  }

  if (Array.isArray(httpConfig)) {
    return await createMultiple(createHttp, httpConfig, log);
  }

  const
    server = require('http').createServer(httpConfig.handler),
    port = httpConfig.port;

  commonPostCreateSetup(httpConfig, server);

  const args = [server, port];
  if (httpConfig.host) {
    args.push(httpConfig.host);
  }

  log('http | try listen ' + port);

  return new Promise((resolve, reject) => {
    args.push(function listener(err) {
      err ? reject(err) : resolve(server);
    });
    connected.apply(null, args);
  });
}

//
// ### function createHttps ()
// Attempts to create and listen on the HTTPS server.
//
async function createHttps(ssl, log, h2) {
  if (typeof ssl === 'undefined') {
    log('https | no options.https; no server');
    return null;
  }

  if (Array.isArray(ssl)) {
    return await createMultiple(createHttps, ssl, log, h2);
  }

  const [key, cert, ca] = await Promise.all([
    normalizePEMContent(ssl.root, ssl.key),
    normalizeCertContent(ssl.root, ssl.cert, ssl.key),
    normalizeCA(ssl.root, ssl.ca)
  ]);

  const finalHttpsOptions = assign({}, ssl, {
    key,
    cert,
    ca,
    //
    // Properly expose ciphers for an A+ SSL rating:
    // https://certsimple.com/blog/a-plus-node-js-ssl
    //
    ciphers: normalizeCiphers(ssl.ciphers),
    honorCipherOrder: !!ssl.honorCipherOrder,
    //
    // Protect against the POODLE attack by disabling SSLv3
    // @see http://googleonlinesecurity.blogspot.nl/2014/10/this-poodle-bites-exploiting-ssl-30.html
    //
    secureProtocol: 'SSLv23_method',
    secureOptions: secureOptions
  });

  if (ssl.sni && !finalHttpsOptions.SNICallback) {
    finalHttpsOptions.SNICallback = await getSNIHandler(ssl);
  }

  const port = ssl.port;
  log('https | listening on %d', port);
  const server = h2
    ? require('http2').createSecureServer(finalHttpsOptions, ssl.handler)
    : require('https').createServer(finalHttpsOptions, ssl.handler);

  commonPostCreateSetup(ssl, server);
  const args = [server, port];
  if (ssl.host) {
    args.push(ssl.host);
  }

  return new Promise((resolve, reject) => {
    args.push(function listener(err) {
      err ? reject(err) : resolve(server);
    });
    connected.apply(null, args);
  });
}

async function createMultiple(createFn, configArray, log) {
  const errorsOrServers = await Promise.allSettled(
    configArray.map(cfg => createFn(cfg, log))
  );

  const errors = [], servers = [];
  for (const result of errorsOrServers) {
    result.reason && errors.push(result.reason);
    result.value && servers.push(result.value);
  }

  if (errors.length) {
    throw errors;
  } else {
    return servers;
  }
}

function commonPostCreateSetup({ timeout, keepAliveTimeout }, server) {
  if (typeof timeout === 'number') {
    server.setTimeout(timeout);
  }
  if (typeof keepAliveTimeout === 'number') {
    server.keepAliveTimeout = keepAliveTimeout;
  }
}
